db/versions/0017.yml

```diff
@@ -0,0 +1,231 @@
+version: 17
+migrationScript: 0017-migration.sql
+downgradeScript: 0017-downgrade.sql
+methods:
+  queue_artifacts_entities_load:
+    description: See taskcluster-lib-entities
+    mode: read
+    serviceName: queue
+    args: partition_key text, row_key text
+    returns: table (partition_key_out text, row_key_out text, value jsonb, version integer, etag uuid)
+    body: |-
+      declare
+        decoded_composite_key text[];
+      begin
+        decoded_composite_key := decode_composite_key(queue_artifacts_entities_load.partition_key);
+        return query
+        select
+          queue_artifacts_entities_load.partition_key,
+          queue_artifacts_entities_load.row_key,
+          entity_buf_encode(
+            jsonb_build_object(
+              'PartitionKey', queue_artifacts_entities_load.partition_key,
+              'RowKey', queue_artifacts_entities_load.row_key,
+              'taskId', slugid_to_uuid(task_id),
+              'runId', run_id,
+              'name', name,
+              'storageType', storage_type,
+              'contentType', content_type,
+              'present', present,
+              'expires', expires),
+            'details', details::text) as value,
+          1 as version,
+          queue_artifacts.etag as etag
+        from queue_artifacts
+        where
+          queue_artifacts.task_id = decoded_composite_key[1] and queue_artifacts.run_id::text = decoded_composite_key[2] and queue_artifacts.name = decode_string_key(queue_artifacts_entities_load.row_key);
```

Comment:
```
postgres=# explain select * from queue_artifacts where task_id='x' and run_id::text = '25';
                                         QUERY PLAN                                          
---------------------------------------------------------------------------------------------
 Index Scan using queue_artifacts_pkey on queue_artifacts  (cost=0.29..8.31 rows=1 width=52)
   Index Cond: (task_id = 'x'::text)
   Filter: ((run_id)::text = '25'::text)
```
In general, any time there's anything but a bare column name in a where clause, there's a risk of a table scan.  In this case, it's not so bad because postgres can use the index to find the task, and then just scan through the runs (of which there are probably 1 and at most 50).  Postgres doesn't know that these are all stringified integers, so it can't translate `run_id::text = key` into `run_id = key::int`.  I think explicitly casting `decoded_composite_key[2]::int` is the fix here (and in all the while clauses).

```
postgres=# explain select * from queue_artifacts where task_id='x' and run_id = '25'::int;
                                         QUERY PLAN                                          
---------------------------------------------------------------------------------------------
 Index Scan using queue_artifacts_pkey on queue_artifacts  (cost=0.29..8.30 rows=1 width=52)
   Index Cond: ((task_id = 'x'::text) AND (run_id = 25))
```

Reply:
That makes a lot of sense. Thank you for both explain queries. It makes it easier to spot in the future.